### PR TITLE
Feat(events): Allow explicit 0 for events to apply directly

### PR DESCRIPTION
**Pull request by RisingLeaf**
**Feature**

This PR addresses the bug/feature described in issue #6719 and as requested by @bene-dictator 

## Summary
Up til now events were only applied on the next increment of date. This PR allows for an explicit time to event of 0, which mean the event gets applied when the game action is triggered.
Take this example:
```html
event "war begins"
	system Sol
		government "Free Worlds"

mission "Name2"
	job
	repeat
	on accept
		event "war begins" 0
```
The map changes will be seen as soon as the mission gets accepted

## Testing Done
I used the test mission to test it once with explicit 0 and without, worked like expected, with explicit 0 map changes were seen directly, without I had to depart

## Performance Impact
N/A

## Side Note
I looked through the data files and there is no place where event gets used with a 0, so no need to change anything there.

### CODE REVIEW

File: source/PlayerInfo.cpp [hunk 1]
```diff
@@ -586,8 +586,19 @@ void PlayerInfo::AddChanges(list<DataNode> &changes)
 // Add an event that will happen at the given date.
 void PlayerInfo::AddEvent(const GameEvent &event, const Date &date)
 {
-	gameEvents.push_back(event);
-	gameEvents.back().SetDate(date);
+	// Check the event should be applied directly.
```
Comment (RisingLeaf) [inline source/PlayerInfo.cpp:589]:
```suggestion
	// Check if the event should be applied directly.
```

File: source/PlayerInfo.cpp [hunk 2]
```diff
@@ -586,8 +586,19 @@ void PlayerInfo::AddChanges(list<DataNode> &changes)
 // Add an event that will happen at the given date.
 void PlayerInfo::AddEvent(const GameEvent &event, const Date &date)
 {
-	gameEvents.push_back(event);
-	gameEvents.back().SetDate(date);
+	// Check if the event should be applied directly.
+	if(date <= this->date)
+	{
+		GameEvent eventCopy = event;
+		list<DataNode> eventChanges = {eventCopy.Apply(*this)};
```
Comment (TomGoodIdea) [inline source/PlayerInfo.cpp:593]: Why copy the event? Btw there's nothing blocking GameEvent::Apply from being const?
```suggestion
		list<DataNode> eventChanges = {event.Apply(*this)};
```

Comment (RisingLeaf) [reply source/PlayerInfo.cpp:593]: it is not const because of the std::move

Comment (RisingLeaf) [reply source/PlayerInfo.cpp:593]: I copy it because it is not const

Comment (TomGoodIdea) [reply source/PlayerInfo.cpp:593]: ah, different std implementations in various compilers, I guess

Comment (RisingLeaf) [reply source/PlayerInfo.cpp:593]: not really, https://en.cppreference.com/w/cpp/utility/move
if you look here you can see that std move does not take in const values

Comment (RisingLeaf) [reply source/PlayerInfo.cpp:593]: std move empties the object it takes and "moves" the data into the other object

**Review by TomGoodIdea — APPROVED** [review]
Tested changing map and player swizzle; normal events still work as they should.
Thank you!

File: source/PlayerInfo.cpp (hunk 2 above)
Comment (TomGoodIdea) [reply source/PlayerInfo.cpp:593]: ~~so why did I manage to compile with const~~

Comment (RisingLeaf) [reply source/PlayerInfo.cpp:593]: thats strange

Comment (RisingLeaf) [reply source/PlayerInfo.cpp:593]: what compiler are you using?

Comment (RisingLeaf) [reply source/PlayerInfo.cpp:593]: ah ok, no I get it now, so std move only tries to move stuff but when its const it just copies it

**Review by bene-dictator — APPROVED** [review]
The single comment in this PR looks grammatically perfect.

Thank you so much :)

File: source/PlayerInfo.cpp (hunk 2 above)
Comment (TomGoodIdea) [reply source/PlayerInfo.cpp:593]: hmm interesting

Comment (TomGoodIdea) [reply source/PlayerInfo.cpp:593]: then it can stay as is

**Comment by bene-dictator** [comment]
Works as expected with the Hai Start PR.

**Review by TomGoodIdea — APPROVED** [review]

**Review by OcelotWalrus — APPROVED** [review]

**Review by tibetiroka — APPROVED** [review]

**Pull request merged**